Teaching the tidyverse in 2021 (#530)
tidyverse/tidyverse.org: content/blog/teach-tidyverse-2021/index.Rmd

---
output: hugodown::hugo_document

slug: teach-tidyverse-2021
title: Teaching the tidyverse in 2021
date: 2021-08-31
author: Mine Çetinkaya-Rundel
description: >

>     Recommendations for teaching the tidyverse in 2021, summarizing package updates most relevant for teaching data science with the tidyverse, particularly to new learners.

Reviewer: Very small, but this needs a line break — since you're using the yaml `>` you can spread across multiple lines with out worrying.

photo:
  url: https://unsplash.com/photos/lj5ALRcon4g
  author: Jackie Hope

# one of: "deep-dive", "learn", "package", "programming", or "other"
categories: [learn] 
tags: [tidyverse, teaching]
---

<!--
TODO:
* [ x ] Look over / edit the post's title in the yaml
* [ x ] Edit (or delete) the description; note this appears in the Twitter card
* [ x ] Pick category and tags (see existing with `hugodown::tidy_show_meta()`)
* [ x ] Find photo & update yaml metadata
* [ ] Create `thumbnail-sq.jpg`; height and width should be equal
* [ ] Create `thumbnail-wd.jpg`; width should be >5x height
* [ ] `hugodown::use_tidy_thumbnails()`
* [ ] Add intro sentence, e.g. the standard tagline for the package

> * [ ] `usethis::use_tidy_thanks()`

Reviewer: Is this up to date?

Author: Oops, it is now!

-->

Last summer I wrote a series of blog posts titled [teaching the tidyverse in 2020](https://education.rstudio.com/blog/2020/07/teaching-the-tidyverse-in-2020-part-4-when-to-purrr/).
As we quickly approach the end of the summer (in the northern hemisphere) and the start of a new academic year, it seems like a good time to provide a new update for teaching the tidyverse, in 2021.
The main audience for this post is educators who teach the tidyverse and who might want to bring their teaching materials up to date with updates to the tidyverse that happened over the past year.
Much of what is discussed here has already been covered in package update posts on this blog, but my goal is to summarize the highlights that are most relevant to teaching data science with the tidyverse, particularly to new learners.

Let's get started!

```{r}
library(tidyverse)
```

> 

Reviewer: Maybe include a brief outline here?

## Lifecycle stages

> The [**lifecycle**](https://lifecycle.r-lib.org/) package is used to manage the lifecycle of functions and features within the tidyverse, with clear messaging about what is still experimental and what the tidyverse team is moving away from in the future.

Reviewer: I feel like this needs a bit more of a lead in — i.e. I think you need a sentence early saying why an educator should care about the lifecycle. Maybe something like:

> The lifecycle stages are a useful guide for teaching because they help you see what the tidyverse is moving forward and what it's moving away from.

The lifecycle stages you might encounter on functions or packages are experimental, stable, deprecated, and superseded.
It's helpful to be aware of the stages (and their associated badges) as you review and revise your teaching materials or as you consider incorporating new tooling into your teaching.

![A diagram showing the transitions between the four main stages: experimental can become stable and stable can become deprecated or superseded.](lifecycle.png)

> -   <img src="lifecycle-stable.svg" alt="Stable" style="vertical-align:middle"/> Stable indicates that breaking changes will be avoided where possible, and they're only made if the long term benefit of such a change exceeds the short term pain of changing existing code.

Reviewer: I think you could emphasise a little more strongly that this is the default state for most functions in the tidyverse.

-   <img src="lifecycle-deprecated.svg" alt="Deprecated" style="vertical-align:middle"/> If a function is noted as deprecated, this means a better alternative is available and this function is scheduled for removal.
    Generally functions will first be soft deprecated and then deprecated.
    Very important functions that become deprecated might next be defunct, which means that function continues to exist but the deprecation warning turns into an error.
    An example of a deprecated function is `tibble::data_frame()`, with the preferred alternative `tibble::tibble()`. Arguments to functions can also be deprecated, e.g., in `tidyr::nest()` the new argument `new_col` makes the former `.key` argument not needed, and hence `.key` is deprecated.
    You should avoid teaching functions that are deprecated and correct their usage in your students' code by suggesting the preferred alternative.

-   <img src="lifecycle-superseded.svg" alt="Superseded" style="vertical-align:middle"/> Superseded indicates that there is a known better alternative for the function, but it's not going away.
    Some examples include the following:

>     -   `tidyr::spread()` / `tidyr::gather()` vs. `tidyr::pivot_longer()` / `tidyr::pivot_wider()`

Reviewer: Link these to blog posts?

    -   Scoped verbs (e.g., `dplyr::mutate_if()`, `dplyr::select_at()`, `dplyr::rename_all()`, etc.) vs. `dplyr::across()`

    -   `dplyr::sample_n()` / `dplyr::sample_frac()` vs. `dplyr::slice_sample()` with `n` and `prop` arguments

    I would recommend not teaching superseded functions to new learners, and for learners who might be aware of them already, I would recommend discouraging their use (though not correcting, i.e., no point deductions on a formative assessment), and suggesting an alternative.

-   <img src="lifecycle-experimental.svg" alt="Experimental" style="vertical-align:middle"/> Experimental functions are made available so the community can try them out and provide feedback, however they come with no promises for long term stability.
    Some examples include the following:

>     -   `dplyr::group_trim()` to drop unused levels of all factors that are used as grouping variables

Reviewer: This is maybe a bit too esoteric?

    -   in `dplyr::summarize()`: `.groups` argument to define the grouping structure of the result

>     -   in `dplyr::mutate()`: `.before` and `.after` arguments to control where new columns should appear

Reviewer: FWIW I'm pretty confident that the experimental badge will be removed from both of these in the next dplyr release. Maybe you could mention that in an aside somehow?

    I would recommend teaching experimental functions with caution, particularly to new learners with whom you might not formally discuss the concept of a "lifecycle".
    However there is no reason to discourage use of these functions -- if students have stumbled upon a solution that involves an experimental function or argument and has used it correctly on their own, this is likely a good indication that the experiment is working!

If you'd like to learn more about the tidyverse lifecycle, I recommend the following resources:

-   Blog post: [lifecycle 1.0.0](https://www.tidyverse.org/blog/2021/02/lifecycle-1-0-0/)
-   Talk: [Maintaining the house the tidyverse built](https://www.rstudio.com/resources/rstudioglobal-2021/maintaining-the-house-the-tidyverse-built/) by Hadley Wickham at rstudio::global(2021)[^1]

[^1]: I think this talk would also be a good resource for software development courses on the topic of maintaining open source software and communicating updates and changes to users.

## Making reproducible examples with reprex

> The [**reprex**](https://reprex.tidyverse.org/) package helps users create **repr**oducible **ex**amples for posting to GitHub issues, StackOverflow, in Slack messages or snippets, or even to paste into PowerPoint or Keynote slides by placing the code to be shared in your clipboard.

Reviewer: Again, I think this needs an early sentence like:

> I find reprex very useful when teaching because it helps my students provide me with broken code in a way that makes it as easy as possible for me to help them.

Writing a reproducible example with a minimal dataset is better practice, but this can be quite difficult for new learners.
Being able to easily use local data will make it easier for them to benefit from other aspects of reprex earlier on.

[^2]: Turns out this was always possible using the `outfile` argument, but now it's more intuitive!

Being able to create a reprex in the current working directory means you can also benefit from a project-level `.Rprofile` if you happen to have one in your project.
This is likely not going to have implications for new learners, for whom this would be an advanced concept, but it can be helpful for instructors who teach with a different suite of packages than what they locally have installed (e.g., CRAN versions for teaching vs. development versions for personal use).
If this describes you, I recommend using [**renv**](https://rstudio.github.io/renv/index.html) in projects where you keep teaching materials, which uses `.Rprofile` to implement a project-specific package library.

> Then, `reprex(wd = ".")` can be used to render using the project-specific library.

Reviewer: ```suggestion
Then, `reprex(wd = ".")` will create a reprex using the packages in that library.
```

For more on updates in reprex, read the blog posts for the [1.0.0](https://www.tidyverse.org/blog/2021/02/reprex-1-0-0/) and [2.0.0](https://www.tidyverse.org/blog/2021/04/reprex-2-0-0/) releases.
And if you're new to reprex, start [here](https://reprex.tidyverse.org/articles/articles/learn-reprex.html).

## Building on tidyverse for modeling with tidymodels

The **tidymodels** framework is a collection of packages for modeling and machine learning using tidyverse principles.
This framework has been around since 2017, but over the past year many of the packages within tidymodels have become stable and gained lots of documentation, making them attractive choices for teaching.
If you're introducing your students to data science with the tidyverse, a great next step to consider is using tidymodels when it comes to modeling and inference.

```{r}
library(tidymodels)
```

From a pedagogical perspective, tidymodels has three main advantages:

1.  Similar interfaces to different models.
2.  Model outputs as tibbles, which are straightforward to interact with for learners who already know how to wrangle and visualize data stored in this format.

> 3.  Features that help users avoid common machine learning pitfalls.

Reviewer: Can you make this more precise? i.e. avoiding over-fitting by making the test-training split a fundamental part of the workflow (except I wouldn't use workflow to avoid confusion with the workflows package).

  set_engine("lm") %>%
  fit(hwy ~ cty, data = mpg) %>%
  tidy()
```

The tidymodels approach takes a few more steps, and for a simple model like this, the only advantage is likely in the summarisation step.
With `tidy()`, we get the model output as a tibble, which is more straightforward to interact with programmatically and which, by default, omits the significant stars.

```{r}
lm(hwy ~ cty, data = mpg) %>%
  tidy()
```

The pedagogical advantages for the consistent API of the framework become more clear when we move on to fitting different models.
Below you can see examples of how we can fit models using various engines or using the same engine, but different modes.

```{r eval = FALSE}
# different engines
linear_reg() %>%
  set_engine("lm") %>%
  set_mode("regression")

logistic_reg() %>% 
  set_engine("glm") %>%
  set_mode("classification")

rand_forest() %>% 
  set_engine("ranger") %>% 
  set_mode("regression")

decision_tree() %>% 
  set_engine("rpart") %>% 
  set_mode("regression")

# same engine, different modes
svm_linear() %>% 
  set_engine("LiblineaR") %>% 
  set_mode("regression")

svm_linear() %>% 
  set_engine("LiblineaR") %>% 
  set_mode("classification")
```

> I should note that fitting a bunch of models to the same data and picking the one you like the results of the best is not a good approach.
> So one would rarely see code as it appears in the chunk above in a single R script.
> Students learning about these models would likely encounter these pipelines over the course of a semester, each in a slightly different data context.

Reviewer: ```suggestion
Fitting a bunch of models to the same data and picking the one you like the results of the best is not a good approach, so one would rarely see code as it appears in the chunk above in a single R script.
Students will encounter these pipelines over the course of a semester, each in a slightly different data context.
Because the syntax is uniform, it's easier to focus on the details of the model, not how to fit the darn thing in R.
```

> Another pedagogical advantage, particularly for teaching tidymodels after tidyverse, is the syntax that resembles dplyr pipelines to build recipes.

Reviewer: ```suggestion
Another pedagogical advantage, particularly for teaching tidymodels after tidyverse, is the syntax to build recipes for feature engineering resembles dplyr pipelines for data wrangling.
```

    month = lubridate::month(time)          # new variable: month
  ) %>%
  select(-time) %>%
  mutate(
    cc = cut(cc, breaks = c(0, 1)),         # discretize cc
    attach = cut(attach, breaks = c(0, 1)), # discretize attach
    dollar = cut(dollar, breaks = c(0, 1))  # discretize dollar
  ) %>%
  mutate(
    inherit = cut(inherit, breaks = c(0, 1, 5, 10, 20)),  # discretize inherit
    password = cut(password, breaks = c(0, 1, 5, 10, 20)) # discretize password
  )

# recipes for data preprocessing and feature engineering
# same steps, similar syntax, 
# less bookkeeping for the analyst in modeling setting
recipe(spam ~ ., data = openintro::email) %>%
  step_rm(from, sent_email) %>%
  step_date(
    time, 
    features = c("dow", "month")
    ) %>%
  step_rm(time) %>%
  step_cut(
    cc, 
    attach, 
    dollar, breaks = c(0, 1)
    ) %>%
  step_cut(
    inherit, 
    password, breaks = c(0, 1, 5, 10, 20)
    )
```

You might be thinking "Why do I need the **recipes** `step_*()` functions when I can express the same steps with dplyr?" This brings us back to the "features that avoid common machine learning pitfalls".

> The advantage of this approach is that once recipe steps are developed with the training data, they can be applied to the testing data with ease.

Reviewer: ```suggestion
The advantage of this approach is that once recipe steps are developed with the training data, they can be automatically applied to the testing data for final model assessment.
```

-   Expanded documentation:

    -   [Get started with tidymodels](https://www.tidymodels.org/start/)
    -   [Learn more and go further](https://www.tidymodels.org/learn/)

-   Book: [Tidy Modeling with R](https://www.tmwr.org/) by Max Kuhn and Julia Silge

-   Blog posts:

    -   [Choose your own tidymodels adventure](Choose%20your%20own%20tidymodels%20adventure)
    -   [infer 1.0.0](https://www.tidyverse.org/blog/2021/08/infer-1-0-0/)

If you're new to teaching tidymodels, the following resources can be helpful:

-   USCOTS 2021 Breakout session: [Tidy up your models](https://bit.ly/tidymodels-uscots21/) (developed and presented with [Debbie Yuster](https://www.ramapo.edu/tas/faculty/debbie-yuster/))
-   [Data Science in a Box](https://datasciencebox.org/making-rigorous-conclusions.html): Slides, application exercises, computing labs, and homework assignments on modelling and inference with tidymodels.

## Reading data with readr

A new version of [**readr**](https://www.tidyverse.org/blog/2021/07/readr-2-0-0/#reading-multiple-files-at-once) was recently released, with lots of updates outlined in [this blog post](https://www.tidyverse.org/blog/2021/07/readr-2-0-0/).

> I think the update that is most relevant to teaching is the new functionality for reading in multiple files at once, or more specifically, reading sets of files with the same columns into one output table in a single command.

Reviewer: ```suggestion
The update most relevant to teaching is the new functionality for reading in multiple files at once, or more specifically, reading sets of files with the same columns into one output table in a single command.
```

```{r}
html %>% html_text() %>% writeLines()
```

Note that the line breaks in the output do not respect the line break defined with `<br>`.

With the new `html_text2()`, `<br>` is handled appropriately and the line breaks follow the expected pattern.

```{r}
html %>% html_text2() %>% writeLines()
```

The output of `html_text2()` is generally what you want, but note that it is slower than `html_text()`.
This might not make a big difference for teaching web scraping as a new topic, but it is worth keeping in mind when the task involves scraping a large amount of data.
Your choice might also depend on what you're going to do next with the data.
For example, if the next step involves tokenizing the scraped text with `tidytext::unnest_tokens()` you might not care how the line breaks were handled in the first step.

Since this change involves the addition of a new function without changing behaviour in any existing functions, incorporating it into your teaching would require testing `html_text2()` in places where you previously used `html_text()` to see if the result is preferable.

> 

Reviewer: I think you can still mention `html_table()` here since it should be generally better even if that one example I provided is wrong 😬. Maybe something like: If `html_table()` didn't work for you in the past, it's worth trying again since it's been rewritten from scratch to more closely match how browsers display tables with merged cells.

## SQL and data.table translations with dbplyr and dtplyr

Two packages that provide interfaces for translations between **dplyr** and SQL and [**data.table**](https://rdatatable.gitlab.io/data.table/) code are **dbplyr** and **dtplyr**.
If you're teaching either of these tools alongside the tidyverse, particularly to students who have learned the tidyverse first, the `show_query()` function can be very helpful for translating tidyverse code into syntaxes used by these tools.

dtplyr translates dplyr pipelines into equivalent data.table code.
To start, we first need to create a `lazy_dt()` object which will record the dplyr actions.
Then, we write a dplyr pipeline as usual and save the result.
The result can be viewed by piping it into `as_tibble()` and the data.table code can be viewed with `show_query()`.

```{r}
library(dtplyr)

mtcars_dt <- lazy_dt(mtcars)

cyl_summary <- mtcars_dt %>% 
  group_by(cyl) %>% 
  summarise(across(disp:wt, mean))

# result
cyl_summary %>% as_tibble()

# query
cyl_summary %>% show_query()
```

> With recent updates, dtplyr can also translate some tidyr functions to data.table as well, e.g., `pivot_wider()`.

Reviewer: ```suggestion
With recent updates, dtplyr can also translate some tidyr functions to data.table, e.g., `pivot_wider()`.
```

fish_encounters_dt <- lazy_dt(fish_encounters)

fish_encounters_wider <- fish_encounters_dt %>%
  pivot_wider(names_from = station, values_from = seen, values_fill = 0)

# result
fish_encounters_wider %>% as_tibble()

# query
fish_encounters_wider %>% show_query()
```

Similarly, dbplyr translates dplyr pipelines into equivalent SQL code.
The only difference in the following example translating tidyr code to SQL code is the function used in the first step, `memdb_frame()`, which creates a database table.

```{r}
library(dbplyr)

fish_encounters_db <- memdb_frame(fish_encounters)

fish_encounters_wider <- fish_encounters_db %>%
  pivot_wider(names_from = station, values_from = seen, values_fill = 0)

# result
fish_encounters_wider %>% as_tibble()

# query
fish_encounters_wider %>% show_query()
```

I recommend the following resources to get started with these packages:

-   [Blog post on dplyr backends](https://www.tidyverse.org/blog/2021/02/dplyr-backends/)
-   [dtplyr translation vignette](https://dtplyr.tidyverse.org/articles/translation.html)
-   [Introduction to dbplyr vignette](https://dbplyr.tidyverse.org/articles/dbplyr.html)

## Other teaching and learning resources

> I'd like to highlight two more teaching and learning resources before I wrap up this post:

Reviewer: I think these should go closer to the top!
